### ISSUE TITLE
v5.19.4 thru blead fails to compile on WinXP / Qt5/MinGW due to missing defines for win32sck.c convert_errno_to_wsa_error()

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13350]
Migrated from [rt.perl.org#120236](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120236) (status was 'resolved')

Searchable as RT120236$

**Comment by p5pRT** [comment]
### From festushagen2002@yahoo.com

Hi all\,

perlbug \-d report at end\.

As the title states\, v5\.19\.4 thru blead fails to compile on WinXP / Qt5/MinGW due to missing defines for win32sck\.c in function convert\_errno\_to\_wsa\_error\(\)

The following errors&#8203;:

gcc \-c  \-I\.\\include \-I\. \-I\.\. \-I\.\.\\lib\\CORE \-DWIN32 \-DPERLDLL \-DPERL\_CORE \-s \-O2 \-DPERL\_IS\_MINIPERL \-omini\\win32sck\.o win32sck\.c
win32sck\.c&#8203;: In function 'convert\_errno\_to\_wsa\_error'&#8203;:
win32sck\.c&#8203;:211&#8203;:10&#8203;: error&#8203;: 'EBADMSG' undeclared \(first use in this function\)
     case EBADMSG&#8203;:
          ^
win32sck\.c&#8203;:211&#8203;:10&#8203;: note&#8203;: each undeclared identifier is reported only once for each function it appears in
win32sck\.c&#8203;:229&#8203;:10&#8203;: error&#8203;: 'EIDRM' undeclared \(first use in this function\)
     case EIDRM&#8203;:
          ^
win32sck\.c&#8203;:247&#8203;:10&#8203;: error&#8203;: 'ENODATA' undeclared \(first use in this function\)
     case ENODATA&#8203;:
          ^
win32sck\.c&#8203;:249&#8203;:10&#8203;: error&#8203;: 'ENOLINK' undeclared \(first use in this function\)
     case ENOLINK&#8203;:
          ^
win32sck\.c&#8203;:251&#8203;:10&#8203;: error&#8203;: 'ENOMSG' undeclared \(first use in this function\)
     case ENOMSG&#8203;:
          ^
win32sck\.c&#8203;:255&#8203;:10&#8203;: error&#8203;: 'ENOSR' undeclared \(first use in this function\)
     case ENOSR&#8203;:
          ^
win32sck\.c&#8203;:257&#8203;:10&#8203;: error&#8203;: 'ENOSTR' undeclared \(first use in this function\)
     case ENOSTR&#8203;:
          ^
win32sck\.c&#8203;:261&#8203;:10&#8203;: error&#8203;: 'ENOTRECOVERABLE' undeclared \(first use in this function\)
     case ENOTRECOVERABLE&#8203;:
          ^
win32sck\.c&#8203;:269&#8203;:10&#8203;: error&#8203;: 'EOTHER' undeclared \(first use in this function\)
     case EOTHER&#8203;:
          ^
win32sck\.c&#8203;:281&#8203;:10&#8203;: error&#8203;: 'ETIME' undeclared \(first use in this function\)
     case ETIME&#8203;:
          ^
win32sck\.c&#8203;:285&#8203;:10&#8203;: error&#8203;: 'ETXTBSY' undeclared \(first use in this function\)
     case ETXTBSY&#8203;:
          ^
win32sck\.c&#8203;: In function 'win32\_select'&#8203;:
win32sck\.c&#8203;:506&#8203;:5&#8203;: warning&#8203;: passing argument 5 of 'select' discards 'const' qualifier from pointer target type \[enabled by default\]
     SOCKET\_TEST\_ERROR\(r = select\(nfds\, &nrd\, &nwr\, &nex\, timeout\)\);
     ^
In file included from c&#8203;:\\qt5\\tools\\mingw48\_32\\i686\-w64\-mingw32\\include\\ws2spi\.h&#8203;:13&#8203;:0\,
                 from win32sck\.c&#8203;:19&#8203;:
c&#8203;:\\qt5\\tools\\mingw48\_32\\i686\-w64\-mingw32\\include\\winsock2\.h&#8203;:986&#8203;:34&#8203;: note&#8203;: expected 'PTIMEVAL' but argument is of type 'const struct timeval \*'
   WINSOCK\_API\_LINKAGE int WSAAPI select\(int nfds\,fd\_set \*readfds\,fd\_set \*writefds\,fd\_set \*exceptfds\,const PTIMEVAL timeout\);
                                  ^
dmake&#8203;:  Warning&#8203;: \-\- Found file corresponding to virtual target \[\.\.\\uudmap\.h\]\.
dmake&#8203;:  Error code 129\, while making 'mini\\win32sck\.o'

\#\#\#\#\#\#\#\#\#\# perlbug \-d output&#8203;:


---
Flags&#8203;:
    category=core
    severity=low


---
Site configuration information for perl 5\.19\.5&#8203;:

Configured by fh at Sun Oct 13 21&#8203;:48&#8203;:45 2013\.

Summary of my perl5 \(revision 5 version 19 subversion 5\) configuration&#8203;:
  Derived from&#8203;: 
  Platform&#8203;:
    osname=MSWin32\, osvers=5\.1\, archname=MSWin32\-x86\-multi\-thread
    uname=''
    config\_args='undef'
    hint=recommended\, useposix=true\, d\_sigaction=undef
    useithreads=define\, usemultiplicity=define
    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
    usemymalloc=n\, bincompat5005=undef
  Compiler&#8203;:
    cc='gcc'\, ccflags =' \-s \-O2 \-DWIN32  \-DPERL\_TEXTMODE\_SCRIPTS \-DPERL\_IMPLICIT\_CONTEXT \-DPERL\_IMPLICIT\_SYS \-DUSE\_PERLIO \-fno\-strict\-aliasing \-mms\-bitfields'\,
    optimize='\-s \-O2'\,
    cppflags='\-DWIN32'
    ccversion=''\, gccversion='4\.8\.0'\, gccosandvers=''
    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='long long'\, lseeksize=8
    alignbytes=8\, prototype=define
  Linker and Libraries&#8203;:
    ld='g\+\+'\, ldflags ='\-s \-L"c&#8203;:\\perl\\lib\\CORE" \-L"C&#8203;:\\Qt5\\Tools\\mingw48\_32\\lib"'
    libpth=C&#8203;:\\Qt5\\Tools\\mingw48\_32\\lib C&#8203;:\\Qt5\\Tools\\mingw48\_32\\i686\-w64\-mingw32\\lib
    libs=\-lmoldname \-lkernel32 \-luser32 \-lgdi32 \-lwinspool \-lcomdlg32 \-ladvapi32 \-lshell32 \-lole32 \-loleaut32 \-lnetapi32 \-luuid \-lws2\_32 \-lmpr \-lwinmm \-lversion \-lodbc32 \-lodbccp32 \-lcomctl32
    perllibs=\-lmoldname \-lkernel32 \-luser32 \-lgdi32 \-lwinspool \-lcomdlg32 \-ladvapi32 \-lshell32 \-lole32 \-loleaut32 \-lnetapi32 \-luuid \-lws2\_32 \-lmpr \-lwinmm \-lversion \-lodbc32 \-lodbccp32 \-lcomctl32
    libc=\, so=dll\, useshrplib=true\, libperl=libperl519\.a
    gnulibc\_version=''
  Dynamic Linking&#8203;:
    dlsrc=dl\_win32\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
    cccdlflags=' '\, lddlflags='\-mdll \-s \-L"c&#8203;:\\perl\\lib\\CORE" \-L"C&#8203;:\\Qt5\\Tools\\mingw48\_32\\lib"'

Locally applied patches&#8203;:
    uncommitted\-changes



---
@&#8203;INC for perl 5\.19\.5&#8203;:
    C&#8203;:/perl/site/lib
    C&#8203;:/perl/lib
    \.



---
Environment for perl 5\.19\.5&#8203;:
    HOME \(unset\)
    LANG \(unset\)
    LANGUAGE \(unset\)
    LD\_LIBRARY\_PATH \(unset\)
    LOGDIR \(unset\)
    PATH=C&#8203;:\\WINDOWS\\system32;C&#8203;:\\WINDOWS;C&#8203;:\\WINDOWS\\System32\\Wbem;C&#8203;:\\Qt5\\5\.1\.1\\mingw48\_32\\bin;C&#8203;:\\Qt5\\Tools\\mingw48\_32\\bin;C&#8203;:\\perl\\bin
    PERL\_BADLANG \(unset\)
    SHELL \(unset\)

\-Enjoy
fh :\)\_~

**Comment by p5pRT** [comment]
### From @steve-m-hay

Thanks for the report\. I will have a look at this\, but I need to get a
gcc\-4\.8\.0 first\. It works fine with 4\.7\.0 and 4\.7\.1\, but I haven't tried
a 4\.8\.0 yet\.

From a quick look at the code\, it must be that
ERRNO\_HAS\_POSIX\_SUPPLEMENT has been defined\, which is likely to mean
that errno\.h provides EADDRINUSE and yet doesn't provide the other
\#defines which have caused the errors\. \(The change in 5\.19\.4 was
intended to fix problems with VC\+\+ 2010 and above\, which now provide
EADDRINUSE and more in errno\.h\.\.\.\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @steve-m-hay

On 16 October 2013 08&#8203;:28\, Steve Hay via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> Thanks for the report\. I will have a look at this\, but I need to get a
> gcc\-4\.8\.0 first\. It works fine with 4\.7\.0 and 4\.7\.1\, but I haven't tried
> a 4\.8\.0 yet\.
>
> From a quick look at the code\, it must be that
> ERRNO\_HAS\_POSIX\_SUPPLEMENT has been defined\, which is likely to mean
> that errno\.h provides EADDRINUSE and yet doesn't provide the other
> \#defines which have caused the errors\. \(The change in 5\.19\.4 was
> intended to fix problems with VC\+\+ 2010 and above\, which now provide
> EADDRINUSE and more in errno\.h\.\.\.\)
>

I've had a quick look at gcc\-4\.8\.1 from www\.mingw\.org\, and there is no
sign of EADDRINUSE in errno\.h or anywhere else\, so I'm puzzled how you
have errors from convert\_errno\_to\_wsa\_error\(\) given that the function
is only present if ERRNO\_HAS\_POSIX\_SUPPLEMENT is defined\, and
win32/include/sys/errno2\.h has&#8203;:

\#ifndef EADDRINUSE        /\* New in VC10 \*/
\#  define EADDRINUSE        WSAEADDRINUSE
\#endif

\#undef ERRNO\_HAS\_POSIX\_SUPPLEMENT
\#if EADDRINUSE \!= WSAEADDRINUSE
\#  define ERRNO\_HAS\_POSIX\_SUPPLEMENT
\#endif

Is there an EADDRINUSE in your MinGW/gcc installation\, and where did
you get it from? I didn't see a 4\.8\.0 available; only a 4\.8\.1 \(at
http&#8203;://sourceforge\.net/projects/mingw/files/MinGW/Base/gcc/Version4/\)\.

**Comment by p5pRT** [comment]
### From @sisyphus



\-\-\-\-\-Original Message\-\-\-\-\- 
From&#8203;: Steve Hay 

> I've had a quick look at gcc\-4\.8\.1 from www\.mingw\.org\, and there is no
> sign of EADDRINUSE in errno\.h or anywhere else

The 4\.8\.1 compiler from mingw64 defines the symbol in errno\.h&#8203;:

\#ifndef EADDRINUSE
\#define EADDRINUSE 100
\#endif

Cheers\,
Rob

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Oct 15 08&#8203;:37&#8203;:50 2013\, festushagen2002@&#8203;yahoo\.com wrote&#8203;:
> Hi all\,
> 
> perlbug \-d report at end\.
> 
> As the title states\, v5\.19\.4 thru blead fails to compile on WinXP /
> Qt5/MinGW due to missing defines for win32sck\.c in function
> convert\_errno\_to\_wsa\_error\(\)
> 

Do you know whether blead compiled on this platform on earlier Perl
releases?  If so\, are you able to conduct a git bisect or otherwise
identify the Perl version in which the failure first appeared?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @steve-m-hay

On 16 October 2013 13&#8203;:23\,  \<sisyphus1@&#8203;optusnet\.com\.au> wrote&#8203;:
>
>
> \-\-\-\-\-Original Message\-\-\-\-\- From&#8203;: Steve Hay
>>
>> I've had a quick look at gcc\-4\.8\.1 from www\.mingw\.org\, and there is no
>> sign of EADDRINUSE in errno\.h or anywhere else
>
>
> The 4\.8\.1 compiler from mingw64 defines the symbol in errno\.h&#8203;:
>
> \#ifndef EADDRINUSE
> \#define EADDRINUSE 100
> \#endif
>

Thanks for that\, I hadn't realized that the mingw\-w64 builds might be
so different\!

I can now reproduce the problem\, and should have it fixed soon\. The
problem arises because this gcc's errno\.h defines EADDRINUSE \(100\)\,
but doesn't define \*all\* of the new \#defines which Microsoft added in
VC\+\+ 2010 \(in the range 100\-140\)\! Some are missing\.\.\. That shouldn't
be too difficult to cope with\, though\.

**Comment by p5pRT** [comment]
### From @steve-m-hay

On 16 October 2013 23&#8203;:39\, James E Keenan via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Tue Oct 15 08&#8203;:37&#8203;:50 2013\, festushagen2002@&#8203;yahoo\.com wrote&#8203;:
>> Hi all\,
>>
>> perlbug \-d report at end\.
>>
>> As the title states\, v5\.19\.4 thru blead fails to compile on WinXP /
>> Qt5/MinGW due to missing defines for win32sck\.c in function
>> convert\_errno\_to\_wsa\_error\(\)
>>
>
> Do you know whether blead compiled on this platform on earlier Perl
> releases?  If so\, are you able to conduct a git bisect or otherwise
> identify the Perl version in which the failure first appeared?
>

The problem is due to the different errno\.h in \(some versions of\)
gcc\-4\.8\.0 causing issues with this commit&#8203;:

http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/ea95436966749b18f0eb3ee17b35e67c949931d0

I should have this fixed for 5\.19\.5\.

**Comment by p5pRT** [comment]
### From @steve-m-hay

Thank you for the report\. This should now be fixed by commit 4f79e9b161\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Thank you for the report\. This should now be fixed by commit 4f79e9b161\.

**Comment by p5pRT** [comment]
@steve-m-hay - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From festushagen2002@yahoo.com

Hiya all\,

Gee\, I guess next time I need to pay closer attention \.\.\. That's AWESOME\!

Sorry so late\, I see it appears to be figured out and headed in the right direction\.

To clarify a bit\.
The MinGW version is the MinGW that comes bundled with Qt5\, built by the MinGW\-builds project\.

C&#8203;:\\Qt5>gcc \-\-version
gcc \(rev2\, Built by MinGW\-builds project\) 4\.8\.0
Copyright \(C\) 2013 Free Software Foundation\, Inc\.
This is free software; see the source for copying conditions\.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE\.

Yup\, it has EADDRINUSE 100 in errno\.h

The last blead I know for sure worked was just prior to v5\.19\.2\, I typically am on blead\, got sidetracked the last couple of months\, got behind\.
 
Yea\, Uh\-Huh\, That would be the commit \.\.\. ea954369

Thanks all for doing what ya do\.

Big Kudos Steve\, I'll be looking for the fix commit\.

\-Enjoy
fh : \)\_~




On Wednesday\, October 16\, 2013 7&#8203;:19 PM\, Steve Hay via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
 
On 16 October 2013 23&#8203;:39\, James E Keenan via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Tue Oct 15 08&#8203;:37&#8203;:50 2013\, festushagen2002@&#8203;yahoo\.com wrote&#8203;:
>> Hi all\,
>>
>> perlbug \-d report at end\.
>>
>> As the title states\, v5\.19\.4 thru blead fails to compile on WinXP /
>> Qt5/MinGW due to missing defines for win32sck\.c in function
>> convert\_errno\_to\_wsa\_error\(\)
>>
>
> Do you know whether blead compiled on this platform on earlier Perl
> releases?  If so\, are you able to conduct a git bisect or otherwise
> identify the Perl version in which the failure first appeared?
>

The problem is due to the different errno\.h in \(some versions of\)
gcc\-4\.8\.0 causing issues with this commit&#8203;:

http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/ea95436966749b18f0eb3ee17b35e67c949931d0

I should have this fixed for 5\.19\.5\.

**Comment by p5pRT** [comment]
### From festushagen2002@yahoo.com

Hi all\,

Hopefully this gets through \.\.\. Previous replys have not\.

All better now\, so far at least\, it's still building\, but it is beyond this issue\.

Thanks to all for doing what ya do\.

Kudos to Steve for a quick response and timely resolution\. Hats off to ya\!

Impressive\!

\-Enjoy
fh : \)\_~




On Thursday\, October 17\, 2013 8&#8203;:57 AM\, Steve Hay via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
 
According to our records\, your request regarding 
  "v5\.19\.4 thru blead fails to compile on WinXP / Qt5/MinGW due to missing defines for win32sck\.c convert\_errno\_to\_wsa\_error\(\)" 
has been resolved\. 

If you have any further questions or concerns\, please respond to this message\.

For other topics\, please create a new ticket\.

\<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120236 >
